crates/evm/core/src/backend/strategy.rs

```diff
@@ -0,0 +1,275 @@
+use std::fmt::Debug;
+
+use crate::InspectorExt;
+
+use super::{BackendInner, DatabaseExt, Fork, ForkDB, ForkType, FoundryEvmInMemoryDB};
+use alloy_primitives::{Address, U256};
+use alloy_rpc_types::serde_helpers::OtherFields;
+use eyre::Context;
+use revm::{
+    db::CacheDB,
+    primitives::{EnvWithHandlerCfg, HashSet, ResultAndState},
+    DatabaseRef, JournaledState,
+};
+use serde::{Deserialize, Serialize};
+
+pub struct BackendStrategyForkInfo<'a> {
+    pub active_fork: Option<&'a Fork>,
+    pub active_type: ForkType,
+    pub target_type: ForkType,
+}
+
+pub trait BackendStrategy: Debug + Send + Sync {
+    fn name(&self) -> &'static str;
+
+    /// When creating or switching forks, we update the AccountInfo of the contract
+    fn update_fork_db(
+        &self,
+        fork_info: BackendStrategyForkInfo<'_>,
+        mem_db: &FoundryEvmInMemoryDB,
+        backend_inner: &BackendInner,
+        active_journaled_state: &mut JournaledState,
+        target_fork: &mut Fork,
+    );
+
+    /// Clones the account data from the `active_journaled_state` into the `fork_journaled_state`
+    fn merge_journaled_state_data(
+        &self,
+        addr: Address,
+        active_journaled_state: &JournaledState,
+        fork_journaled_state: &mut JournaledState,
+    );
+
+    fn merge_db_account_data(&self, addr: Address, active: &ForkDB, fork_db: &mut ForkDB);
+
+    fn set_inspect_context(&mut self, other_fields: OtherFields);
+
+    /// Executes the configured test call of the `env` without committing state changes.
+    ///
+    /// Note: in case there are any cheatcodes executed that modify the environment, this will
+    /// update the given `env` with the new values.
+    #[instrument(name = "inspect", level = "debug", skip_all)]
+    fn call_inspect<'i, 'db>(
+        &mut self,
+        db: &'db mut dyn DatabaseExt,
+        env: &mut EnvWithHandlerCfg,
+        inspector: &'i mut dyn InspectorExt,
+    ) -> eyre::Result<ResultAndState> {
+        let mut evm = crate::utils::new_evm_with_inspector(db, env.clone(), inspector);
+
+        let res = evm.transact().wrap_err("backend: failed while inspecting")?;
+
+        env.env = evm.context.evm.inner.env;
+
+        Ok(res)
+    }
+
+    /// Executes the configured test call of the `env` without committing state changes.
+    ///
+    /// Note: in case there are any cheatcodes executed that modify the environment, this will
+    /// update the given `env` with the new values.
+    #[instrument(name = "inspect", level = "debug", skip_all)]
+    fn transact_inspect<'i, 'db>(
```

Comment:
Maybe there's a good explanation between the difference between `call` and `transact`. Seems in this one you want you might want to use another environment, and here `call` would mean a `CALL` opcode and `transact` would mean a full transaction to the network. If so, we would have to also deal with the ambiguity that foundry uses `call` to mean run a tx without commiting changes and `transact` means do the same while comitting, at least for the [executor](https://github.com/foundry-rs/foundry/blob/dabacecdc14d074a108c18f97d1e1f63ade37a37/crates/evm/evm/src/executors/mod.rs#L66).

Reply:
FWIW I still don't understand the semantic difference between call and transact, seems an implementation detail to just inject another env, and unrelated to how call/transact are used throughout foundry.